### PR TITLE
cmd/snap: make the buy tests run again and fix the failures

### PR DESCRIPTION
**Pull request by pete-woods**

### CODE REVIEW

**Comment by pete-woods** [comment]
Interestingly, the osutil.FileExists function I just switched to does the comparison on err, too.

**Comment by niemeyer** [comment]
@chipaca's point is that if you remove the file and it returns nil, either it worked or the kernel has a serious bug, so we tend to trust this as evidence.

On the other hand, if FileExists is removing the file, we have a serious bug on our end! :-)

**Comment by pete-woods** [comment]
Misread the comment!

File: cmd/snap/main_test.go [hunk 1]
```diff
@@ -93,10 +93,11 @@ func (s *BaseSnapSuite) Login(c *C) {
...
 
 func (s *BaseSnapSuite) Logout(c *C) {
-	c.Assert(s.AuthFile == "", Equals, false)
-	err := os.Remove(s.AuthFile)
-	c.Assert(err, IsNil)
-	c.Assert(osutil.FileExists(s.AuthFile), Equals, false)
+	if osutil.FileExists(s.AuthFile) {
+		err := os.Remove(s.AuthFile)
+		c.Assert(err, IsNil)
+		c.Assert(osutil.FileExists(s.AuthFile), Equals, false)
```
Comment (niemeyer) [inline cmd/snap/main_test.go]: These three lines could be simply:
```
c.Assert(os.Remove(s.AuthFile), IsNil)
```
If Remove worked, we usually trust it did its job.

**Comment by niemeyer** [comment]
@chipaca as a nitpick I'm clarifying mainly because I've seen code like this before, so for the benefit of Future Generations:

> osutil.FileExists and os.Stat returning nil error are the same thing

That's not quite true, in the sense that os.Stat _failing_ does not imply the file _not_ existing. The equivalent check would be !os.IsNotExist(statErr).

File: cmd/snap/main_test.go (hunk 1 above)
Comment (pete-woods) [inline cmd/snap/main_test.go]: Made this change

**Comment by chipaca** [comment]
> @chipaca's point is that if you remove the file and it returns nil, either it worked or the kernel has a serious bug, so we tend to trust this as evidence.

@niemeyer you give me more credit than is due. I was saying "this code smells". You pointed out the rot behind the smell. Huzzah.

> @chipaca as a nitpick I'm clarifying mainly because I've seen code like this before, so for the benefit of Future Generations:
> > osutil.FileExists and os.Stat returning nil error are the same thing
> That's not quite true, in the sense that os.Stat failing does not imply the file not existing. The equivalent check would be !os.IsNotExist(statErr).

I agree with you in theory, but in practice I was exactly correct in what I was saying:

```go
// FileExists return true if given path can be stat()ed by us. Note that
// it may return false on e.g. permission issues.
func FileExists(path string) bool {
	_, err := os.Stat(path)
	return err == nil
}
```

this has bugged me a few times, but never enough to change it to check `os.IsNotExist` (because when I tried to do that tests blew up IIRC, and last time I noticed we were once again in a crunch).

**Comment by chipaca** [comment]
@niemeyer I can confirm that changing `osutil.FileExists` from its current meaning of "we can stat the path provided" to "we know the path provided exists" breaks a lot of things in strange ways. E.g,
```
FAIL: kernel_os_test.go:65: kernelOSSuite.TestExtractKernelAssetsAndRemove

kernel_os_test.go:84:
    c.Assert(err, IsNil)
... value *os.PathError = &os.PathError{Op:"open", Path:"/tmp/check-9193736347556640924/2/snapsrc/ubuntu-kernel_4.0-1_all.snap/meta/snap.yaml", Err:0x14} ("open /tmp/check-9193736347556640924/2/snapsrc/ubuntu-kernel_4.0-1_all.snap/meta/snap.yaml: not a directory")
```

**Comment by niemeyer** [comment]
Ouch!

**Pull request merged**